### PR TITLE
Adding new features to runtest.py

### PR DESCRIPTION
**Pull request by fabamatic**
|Related issue|
|---|
|#13555|

## Description

New features added to runtest.py:
* Windows event channel testing
* Better reporting capabilities

## Tests

|Component |  Tested  |  Total   | Coverage |
| -------- | -------- | -------- | -------- |
|  Rules   |   1171   |   4247   |  27.57%  |
| Decoders |    75    |   165    |  45.45%  |

|          File           |  Passed  |  Failed  |  Status  |
|        --------         | -------- | -------- | -------- |
|./tests/proftpd.ini      |    7     |    0     |   ✅    |
|./tests/exim.ini         |    7     |    0     |   ✅    |
|./tests/squid_rules.ini  |    2     |    0     |   ✅    |
|./tests/sysmon_eid_3.ini |    10    |    0     |   ✅    |
|./tests/checkpoint_smart1.ini|    18    |    0     |   ✅    |
|./tests/iptables.ini     |    8     |    0     |   ✅    |
|./tests/sysmon_eid_10.ini|    4     |    0     |   ✅    |
|./tests/SonicWall.ini    |    11    |    0     |   ✅    |
|./tests/panda_paps.ini   |    8     |    0     |   ✅    |
|./tests/fortiauth.ini    |    4     |    0     |   ✅    |
|./tests/openldap.ini     |    9     |    0     |   ✅    |
|./tests/f5_big_ip.ini    |    48    |    0     |   ✅    |
|./tests/sophos.ini       |    8     |    0     |   ✅    |
|./tests/opensmtpd.ini    |    7     |    0     |   ✅    |
|./tests/netscreen.ini    |    4     |    0     |   ✅    |
|./tests/rsh.ini          |    2     |    0     |   ✅    |
|./tests/arbor.ini        |    2     |    0     |   ✅    |
|./tests/web_rules.ini    |    10    |    0     |   ✅    |
|./tests/exchange.ini     |    2     |    0     |   ✅    |
|./tests/vuln_detector.ini|    2     |    0     |   ✅    |
|./tests/sysmon_eid_7.ini |    6     |    0     |   ✅    |
|./tests/cpanel.ini       |    7     |    0     |   ✅    |
|./tests/apparmor.ini     |    5     |    0     |   ✅    |
|./tests/pam.ini          |    5     |    0     |   ✅    |
|./tests/apache.ini       |    12    |    0     |   ✅    |
|./tests/fireeye.ini      |    3     |    0     |   ✅    |
|./tests/sysmon.ini       |    3     |    0     |   ✅    |
|./tests/dovecot.ini      |    15    |    0     |   ✅    |
|./tests/nginx.ini        |    12    |    0     |   ✅    |
|./tests/nextcloud.ini    |    8     |    0     |   ✅    |
|./tests/unbound.ini      |    0     |    0     |   ✅    |
|./tests/paloalto.ini     |    16    |    0     |   ✅    |
|./tests/sysmon_eid_13.ini|    7     |    0     |   ✅    |
|./tests/ossec.ini        |    5     |    0     |   ✅    |
|./tests/cisco_ftd.ini    |    42    |    0     |   ✅    |
|./tests/sysmon_eid_1.ini |    59    |    0     |   ✅    |
|./tests/github.ini       |   324    |    0     |   ✅    |
|./tests/fortimail.ini    |    6     |    0     |   ✅    |
|./tests/named.ini        |    5     |    0     |   ✅    |
|./tests/powershell.ini   |    32    |    0     |   ✅    |
|./tests/openvpn_ldap.ini |    2     |    0     |   ✅    |
|./tests/sysmon_eid_8.ini |    4     |    0     |   ✅    |
|./tests/cloudflare-waf.ini|    13    |    0     |   ✅    |
|./tests/huawei_usg.ini   |    3     |    0     |   ✅    |
|./tests/eset.ini         |    8     |    0     |   ✅    |
|./tests/gcp.ini          |    31    |    0     |   ✅    |
|./tests/samba.ini        |    4     |    0     |   ✅    |
|./tests/pfsense.ini      |    2     |    0     |   ✅    |
|./tests/cisco_asa.ini    |    88    |    0     |   ✅    |
|./tests/systemd.ini      |    2     |    0     |   ✅    |
|./tests/web_appsec.ini   |    31    |    0     |   ✅    |
|./tests/sysmon_eid_11.ini|    28    |    0     |   ✅    |
|./tests/cisco_ios.ini    |    17    |    0     |   ✅    |
|./tests/pix.ini          |    22    |    0     |   ✅    |
|./tests/php.ini          |    2     |    0     |   ✅    |
|./tests/office365.ini    |   128    |    0     |   ✅    |
|./tests/fortigate.ini    |    45    |    0     |   ✅    |
|./tests/audit_scp.ini    |    8     |    0     |   ✅    |
|./tests/win_application.ini|    0     |    0     |   ✅    |
|./tests/fortiddos.ini    |    1     |    0     |   ✅    |
|./tests/cimserver.ini    |    2     |    0     |   ✅    |
|./tests/freepbx.ini      |    6     |    0     |   ✅    |
|./tests/sshd.ini         |    47    |    0     |   ✅    |
|./tests/win_event_channel.ini|    8     |    0     |   ✅    |
|./tests/gitlab.ini       |    27    |    0     |   ✅    |
|./tests/glpi.ini         |    3     |    0     |   ✅    |
|./tests/api.ini          |    20    |    0     |   ✅    |
|./tests/dropbear.ini     |    3     |    0     |   ✅    |
|./tests/firewalld.ini    |    2     |    0     |   ✅    |
|./tests/mailscanner.ini  |    1     |    0     |   ✅    |
|./tests/owlh.ini         |    4     |    0     |   ✅    |
|./tests/sysmon_eid_20.ini|    2     |    0     |   ✅    |
|./tests/oscap.ini        |    32    |    0     |   ✅    |
|./tests/mcafee_epo.ini   |    1     |    0     |   ✅    |
|./tests/doas.ini         |    4     |    0     |   ✅    |
|./tests/junos.ini        |    3     |    0     |   ✅    |
|./tests/sudo.ini         |    8     |    0     |   ✅    |
|./tests/syslog.ini       |    6     |    0     |   ✅    |
|./tests/sophos_fw.ini    |    10    |    0     |   ✅    |
|./tests/vsftpd.ini       |    4     |    0     |   ✅    |
|./tests/postfix.ini      |    2     |    0     |   ✅    |
|./tests/modsecurity.ini  |    6     |    0     |   ✅    |
|./tests/su.ini           |    5     |    0     |   ✅    |
|./tests/kernel_usb.ini   |    6     |    0     |   ✅    |
|./tests/aws_s3_access.ini|    10    |    0     |   ✅    |
|./tests/auditd.ini       |    31    |    0     |   ✅    |

### CODE REVIEW

File: ruleset/testing/runtests.py [hunk 1]
```diff
@@ -147,12 +209,46 @@ def run(self, selective_test=False, geoip=False, custom=False):
                                 neg = True
                             else:
                                 neg = False
-                            self.runTest(value, rule, alert, decoder,
-                                         t, name, negate=neg)
+                            self._execution_data[aFile][self.runTest(value, rule, alert, decoder,t, name, negate=neg)]+=1
                 print("")
                 print("")
```
Comment (jmv74211) [inline ruleset/testing/runtests.py:214]: ?

File: ruleset/testing/runtests.py [hunk 2]
```diff
@@ -181,11 +277,32 @@ def cleanup(*args):
...
+    rules = getRuleIds("/var/ossec/ruleset/rules/")
+    decoders = getParentDecoderNames("/var/ossec/ruleset/decoders/")
+
+    template = "|{: ^10}|{: ^10}|{: ^10}|{: ^10}|"
+    print(template.format("Component", "Tested", "Total", "Coverage"))
+    print(template.format("--------", "--------", "--------", "--------"))
+    template = "|{: ^10}|{: ^10}|{: ^10}|{: ^10.2%}|"
+    print(template.format("Rules", len(OT.tested_rules), len(rules), len(OT.tested_rules)/len(rules)))
+    print(template.format("Decoders", len(OT.tested_decoders), len(decoders), len(OT.tested_decoders)/len(decoders)))
+    print("")
```
Comment (jmv74211) [inline ruleset/testing/runtests.py:303]: ?

File: ruleset/testing/runtests.py [hunk 3]
```diff
@@ -181,11 +277,32 @@ def cleanup(*args):
...
     provisionDR()
     restart_analysisd()
+
     OT = OssecTester(args.wazuh_home)
-    error = OT.run(selective_test, args.geoip, args.custom)
+    error = OT.run(selective_test, args.geoip)
+
     cleanDR()
+    if args.windows_tests:
+            disable_win_eventlog_tests()
```
Comment (jmv74211) [inline ruleset/testing/runtests.py:291]: Bad indentation

**Review by jmv74211 — CHANGES_REQUESTED** [review]

**Review by jmv74211 — COMMENTED** [review]
Review of the format and structure of the code.

Main problems encountered:

- `snake_case` format is not used in `coverage.py`
- Inconsistency of styles between different files
- Sometimes descriptive variable names are not used.
- Undocumented code and methods. Difficult to maintain code.
- Linter reports many formatting errors:
  - `coverage.py`

    ```
    coverage.py:16:1: E302 expected 2 blank lines, found 1
    coverage.py:39:1: E302 expected 2 blank lines, found 1
    coverage.py:69:41: E231 missing whitespace after ','
    coverage.py:70:32: E201 whitespace after '('
    coverage.py:70:99: E202 whitespace before ')'
    coverage.py:74:1: E302 expected 2 blank lines, found 1
    coverage.py:105:44: E231 missing whitespace after ','
    coverage.py:135:1: E305 expected 2 blank lines after class or function definition, found 1
    coverage.py:147:1: W391 blank line at end of file
    ```

  - `runtests.py`

    ```
    runtests.py:39:5: E722 do not use bare 'except'
    runtests.py:51:1: E302 expected 2 blank lines, found 1
    runtests.py:58:80: E231 missing whitespace after ','
    runtests.py:59:39: E203 whitespace before ','
    runtests.py:63:83: E231 missing whitespace after ','
    runtests.py:64:39: E203 whitespace before ','
    runtests.py:66:1: E302 expected 2 blank lines, found 1
    runtests.py:72:80: E231 missing whitespace after ','
    runtests.py:77:83: E231 missing whitespace after ','
    runtests.py:80:1: E302 expected 2 blank lines, found 1
    runtests.py:84:1: E302 expected 2 blank lines, found 1
    runtests.py:88:1: E302 expected 2 blank lines, found 1
    runtests.py:93:41: E231 missing whitespace after ','
    runtests.py:97:1: E302 expected 2 blank lines, found 1
    runtests.py:101:41: E231 missing whitespace after ','
    runtests.py:103:39: E231 missing whitespace after ','
    runtests.py:107:1: E302 expected 2 blank lines, found 1
    runtests.py:108:36: E203 whitespace before ':'
    runtests.py:109:21: E127 continuation line over-indented for visual indent
    runtests.py:109:36: E203 whitespace before ':'
    runtests.py:110:39: E203 whitespace before ':'
    runtests.py:111:30: E203 whitespace before ':'
    runtests.py:112:32: E203 whitespace before ':'
    runtests.py:120:91: E231 missing whitespace after ','
    runtests.py:123:63: E231 missing whitespace after ','
    runtests.py:124:121: E501 line too long (174 > 120 characters)
    runtests.py:124:166: E231 missing whitespace after ','
    runtests.py:135:1: E302 expected 2 blank lines, found 1
    runtests.py:193:56: E231 missing whitespace after ':'
    runtests.py:193:68: E231 missing whitespace after ':'
    runtests.py:212:97: E231 missing whitespace after ','
    runtests.py:212:119: E225 missing whitespace around operator
    runtests.py:212:121: E501 line too long (121 > 120 characters)
    runtests.py:236:121: E501 line too long (155 > 120 characters)
    runtests.py:238:121: E501 line too long (136 > 120 characters)
    runtests.py:240:121: E501 line too long (151 > 120 characters)
    runtests.py:244:34: E225 missing whitespace around operator
    runtests.py:245:41: E231 missing whitespace after ','
    runtests.py:245:52: E231 missing whitespace after ','
    runtests.py:246:47: E231 missing whitespace after ','
    runtests.py:246:56: E231 missing whitespace after ','
    runtests.py:256:1: E305 expected 2 blank lines after class or function definition, found 1
    runtests.py:265:121: E501 line too long (129 > 120 characters)
    runtests.py:291:13: E117 over-indented
    ```

File: ruleset/testing/coverage.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    print(len(ruleSet))
+    print(len(parentDecoderSet))
+
+if __name__ == "__main__":
+    # Console arguments definition
+    parser = argparse.ArgumentParser(
+        prog="mitre_mapping.py",
+        description='Script to update ruleset mitre mappings from csv file')
+
+    parser.add_argument("tests_path", help="path were ini test files are located")
```
[the remark this answers is not in the document]
Comment (fabamatic) [reply ruleset/testing/coverage.py:141]: Will remove main method as it's not being used by any scripts

File: ruleset/testing/runtests.py [hunk 5]
```diff
@@ -15,9 +16,11 @@
 import argparse
 import re
 import signal
+import xml.etree.ElementTree as ET
+
+from coverage import getRuleIds
+from coverage import getParentDecoderNames
```
[the remark this answers is not in the document]
Comment (fabamatic) [reply ruleset/testing/runtests.py:22]: Done

File: ruleset/testing/runtests.py [hunk 6]
```diff
@@ -66,39 +69,97 @@ def cleanDR():
...
+
+        if re.search(r"Phase 3: Completed filtering \(rules\)",std_out):
+            rule_search = re.search(r"Completed filtering \(rules\)\.\n\tid:\s+\'(?P<rule_id>\d+)\W+level:\s*'(?P<level>\d+)\W+description:\s+'(?P<description>.*?)'",std_out)
+            failed_test["actual_rule"] = rule_search.group("rule_id")
+            failed_test["actual_level"] = rule_search.group("level")
+            failed_test["description"] = rule_search.group("description")
+        else:
+            failed_test["actual_rule"] = ""
+            failed_test["actual_level"] = ""
+            failed_test["description"] = ""
```
[the remark this answers is not in the document]
Comment (fabamatic) [reply ruleset/testing/runtests.py:138]: done

File: ruleset/testing/runtests.py [hunk 7]
```diff
@@ -66,39 +69,97 @@ def cleanDR():
...
+        self.tested_decoders = set()
 
     def buildCmd(self, rule, alert, decoder):
         cmd = ['%s/wazuh-logtest' % (self._ossec_path), ]
-        cmd += ['-q']
         cmd += ['-U', "%s:%s:%s" % (rule, alert, decoder)]
         return cmd
 
     def runTest(self, log, rule, alert, decoder, section, name, negate=False):
+        did_pass = "failed"
```
[the remark this answers is not in the document]
Comment (fabamatic) [reply ruleset/testing/runtests.py:156]: changed variable name to test_status

File: ruleset/testing/runtests.py [hunk 8]
```diff
@@ -111,24 +172,25 @@ def runTest(self, log, rule, alert, decoder, section, name, negate=False):
...
         for aFile in os.listdir(self._test_path):
-            if re.match(r'^test_(.*?).ini$',aFile) and not custom:
-                continue
             aFile = os.path.join(self._test_path, aFile)
             if aFile.endswith(".ini"):
                 if selective_test and not aFile.endswith(selective_test):
                     continue
                 if geoip is False and aFile.endswith("geoip.ini"):
                     continue
+                self._execution_data[aFile] = {"passed":0, "failed":0}
```
[the remark this answers is not in the document]
Comment (fabamatic) [reply ruleset/testing/runtests.py:196]: Legacy varable name, should be something like "hopefully_a_ini_file", will change to "a_ini_file"

File: ruleset/testing/runtests.py (hunk 3 above)
Comment (fabamatic) [reply ruleset/testing/runtests.py:294]: Corrected

File: ruleset/testing/runtests.py (hunk 2 above)
Comment (fabamatic) [reply ruleset/testing/runtests.py:306]: change to print("\n")

**Review by jmv74211 — APPROVED** [review]

**Review by 72nomada — APPROVED** [review]

**Pull request merged**